File: dotnet/src/extensions/SKHandleBars/Planners/HandlebarsPlanner.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Reflection;
using System.Text.RegularExpressions;
using Microsoft.SemanticKernel.AI;
using Microsoft.SemanticKernel.Connectors.AI.OpenAI.AzureSdk;
using Microsoft.SemanticKernel.Orchestration;

namespace Microsoft.SemanticKernel.Handlebars;

public sealed class HandlebarsPlanner
{
    private readonly IKernel Kernel;

    private readonly HandlebarsPlannerConfiguration Configuration;

    public HandlebarsPlanner(IKernel kernel, HandlebarsPlannerConfiguration? configuration = default)
    {
        this.Kernel = kernel;
        this.Configuration = configuration ?? new HandlebarsPlannerConfiguration();
    }
    public async Task<HandlebarsPlan> CreatePlanAsync(string goal, CancellationToken cancellationToken = default)
    {
        string plannerTemplate;

        Assembly assembly = Assembly.GetExecutingAssembly();
        using (Stream stream = assembly.GetManifestResourceStream("HandlebarPlanner.prompt.yaml")!)
        using (StreamReader reader = new(stream))
        {
            plannerTemplate = reader.ReadToEnd();
        }

        var plannerFunction = SemanticFunction.GetFunctionFromYamlContent(
            plannerTemplate,
            cancellationToken: cancellationToken
        );

        // Get functions
        var functions = ((Kernel)this.Kernel).GetFunctionViews().Where(f => 
        {
            var fullyQualifiedName = $"{f.PluginName}.{f.Name}";
            bool shouldInclude = false;
            if (Configuration.IncludedPlugins.Count == 0 && Configuration.IncludedFunctions.Count == 0)
            {
                shouldInclude = true;
            }
            
            if (Configuration.IncludedPlugins.Contains(f.PluginName))
            {
                shouldInclude = true;
            }
            if (Configuration.IncludedFunctions.Contains(fullyQualifiedName))
            {
                shouldInclude = true;
            }
            if (Configuration.ExcludedPlugins.Contains(f.PluginName))
            {
                shouldInclude = false;
            }
            if (Configuration.ExcludedFunctions.Contains(fullyQualifiedName))
            {
                shouldInclude = false;
            }

            return shouldInclude;
        }).ToList();

        // Generate the plan
        var result = await this.Kernel.RunAsync(
            plannerFunction,
            variables: new Dictionary<string, object?>()
            {
                { "functions", functions},
                { "goal", goal },
                { "lastPlan", Configuration.LastPlan },
                { "lastError", Configuration.LastError }
            }
        );

        List<string> templates = new();
        result.TryGetMetadataValue(AIFunctionResultExtensions.ModelResultsMetadataKey, out IEnumerable<ModelResult> results);
        foreach(ModelResult modelResult in results)
        {
            string template = modelResult.GetResult<ChatModelResult>().Choice.Message.Content!;
            
            // Loop over all the functions and replace "." with "_"
            foreach (var function in functions)
            {
                // Replace without case sensitivity
                template = Regex.Replace(template, $"{function.PluginName}.{function.Name}", $"{function.PluginName}_{function.Name}", RegexOptions.IgnoreCase);
            }

            template = template.Replace($"compare.equal", $"equal");
            template = template.Replace($"compare.lessThan", $"lessThan");
            template = template.Replace($"compare.greaterThan", $"greaterThan");
            template = template.Replace($"compare.lessThanOrEqual", $"lessThanOrEqual");
            template = template.Replace($"compare.greaterThanOrEqual", $"greaterThanOrEqual");
            template = template.Replace($"compare.greaterThanOrEqual", $"greaterThanOrEqual");

            template = MinifyHandlebarsTemplate(template);

            // template = template.Replace("#set", "set");
            // template = template.Replace("{{Guess:", "{{! Guess:");

            templates.Add(template);
        }

        return new HandlebarsPlan(Kernel, templates);
    }
    static string MinifyHandlebarsTemplate(string template)
    {
        // This regex pattern matches '{{', then any characters including newlines (non-greedy), then '}}'
        string pattern = @"(\{\{[\s\S]*?}})";

        // Replace all occurrences of the pattern in the input template
        return Regex.Replace(template, pattern, m =>
        {
            // For each match, remove the whitespace within the handlebars, except for spaces
            // that separate different items (e.g., 'json' and '(get')
            return Regex.Replace(m.Value, @"\s+", " ").Replace(" {", "{").Replace(" }", "}").Replace(" )", ")");
        });
    }
}

public class HandlebarsPlannerConfiguration
{
    public HandlebarsPlannerConfiguration(
        List<string>? includedPlugins = default,
        List<string>? excludedPlugins = default,
        List<string>? includedFunctions = default,
        List<string>? excludedFunctions = default,
        HandlebarsPlan? lastPlan = default,
        string? lastError = default
    )
    {
        IncludedPlugins = includedPlugins ?? new List<string>();
        ExcludedPlugins = excludedPlugins ?? new List<string>();
        IncludedFunctions = includedFunctions ?? new List<string>();
        ExcludedFunctions = excludedFunctions ?? new List<string>();
        LastPlan = lastPlan;
        LastError = lastError;
    }
    public List<string> IncludedPlugins { get; set; }
    public List<string> ExcludedPlugins { get; set; }
    public List<string> IncludedFunctions { get; set; }
    public List<string> ExcludedFunctions { get; set; }
    public HandlebarsPlan? LastPlan { get; set; }
    public string? LastError { get; set; }
}

File: python/src/planners/handlebars_planner.py
import os
from typing import Any

from pydantic import Field
from semantic_kernel.sk_pydantic import SKBaseModel

from ..kernel import newKernel
from ..plugins.semantic_function import SemanticFunction
from .handlebars_plan import HandleBarsPlan


class HandleBarsPlannerConfig(SKBaseModel):
    excluded_functions: list[str] = []
    excluded_plugins: list[str] = []
    included_functions: list[str] = []
    included_plugins: list[str] = []
    last_plan: str | None = None
    last_error: str | None = None


class HandleBarsPlanner(SKBaseModel):
    kernel: newKernel
    configuration: HandleBarsPlannerConfig = Field(
        default_factory=HandleBarsPlannerConfig
    )

    async def create_plan(
        self, goal: str, variables: dict[str, Any] | None = None
    ) -> HandleBarsPlan:
        # read template
        planner_function = SemanticFunction.from_path(
            path=os.getcwd() + "/python/src/planners/handlebar_planner.prompt.yaml"
        )
        functions = [
            function
            for plugin in self.kernel.plugins
            for func_name, function in plugin.fqn_functions.items()
            if self._should_include_function(plugin.name, func_name)
        ]
        plan = await self.kernel.run_async(
            planner_function,
            variables={
                "functions": functions,
                "goal": goal,
                "last_plan": self.configuration.last_plan,
                "last_error": self.configuration.last_error,
            },
            request_settings={"stream": False},
        )
        return HandleBarsPlan(kernel=self.kernel, template=plan["result"])

    def _should_include_function(self, plugin: str, function: str) -> bool:
        """Return True if the function will be included."""
        should_include = (
            len(self.configuration.included_plugins) == 0
            and len(self.configuration.included_functions) == 0
        )
        if plugin in self.configuration.included_plugins:
            should_include = True
        if function in self.configuration.included_functions:
            should_include = True
        if plugin in self.configuration.excluded_plugins:
            should_include = False
        if function in self.configuration.excluded_functions:
            should_include = False
        return should_include
